### ISSUE TITLE
Immediate buf panic in protoreflect with duplicate package names in subdirs

### ISSUE THREAD

**Issue by nashikb** [bufbuild/buf#1071]
When running buf with bare minimum `buf.yaml` and `buf.gen.yaml` files, it immediately crashes per the below.

Given that our production workflows are dead in the water, and that jhump/protoreflect has other unaddressed crashing issues, this makes me incrEDibly nervous about continuing to use buf, as convenient as it is.

```
% buf generate -v
Failure: panic: runtime error: invalid memory address or nil pointer dereference, stack:
goroutine 11 [running]:
runtime/debug.Stack()
	runtime/debug/stack.go:24 +0x68
github.com/bufbuild/buf/private/bufpkg/bufimage/bufimagebuild.getBuildResults.func1.2()
	github.com/bufbuild/buf/private/bufpkg/bufimage/bufimagebuild/builder.go:158 +0x40
panic({0x1035288a0, 0x103cb6c40})
	runtime/panic.go:838 +0x204
github.com/jhump/protoreflect/desc/protoparse.(*linker).createDescriptorPool(0x140010d07d0)
	github.com/jhump/protoreflect@v1.11.1-0.20220213155251-0c2aedc66cf4/desc/protoparse/linker.go:149 +0x4b4
github.com/jhump/protoreflect/desc/protoparse.(*linker).linkFiles(0x140010d07d0)
	github.com/jhump/protoreflect@v1.11.1-0.20220213155251-0c2aedc66cf4/desc/protoparse/linker.go:35 +0x24
github.com/jhump/protoreflect/desc/protoparse.Parser.ParseFiles({{0x0, 0x0, 0x0}, 0x0, 0x0, 0x0, 0x1400000e9c0, 0x1, 0x0, 0x0, ...}, ...)
	github.com/jhump/protoreflect@v1.11.1-0.20220213155251-0c2aedc66cf4/desc/protoparse/parser.go:232 +0x288
github.com/bufbuild/buf/private/bufpkg/bufimage/bufimagebuild.getBuildResult({0x10367a740, 0x1400003cab0}, {0x10367bc30?, 0x1400069e960}, {0x14000600ac0?, 0x7, 0x7}, 0x0)
	github.com/bufbuild/buf/private/bufpkg/bufimage/bufimagebuild/builder.go:211 +0x1ec
github.com/bufbuild/buf/private/bufpkg/bufimage/bufimagebuild.getBuildResults.func1()
	github.com/bufbuild/buf/private/bufpkg/bufimage/bufimagebuild/builder.go:162 +0xb0
created by github.com/bufbuild/buf/private/bufpkg/bufimage/bufimagebuild.getBuildResults
	github.com/bufbuild/buf/private/bufpkg/bufimage/bufimagebuild/builder.go:141 +0x664
```

```
% uname -a    
Darwin osmium.marin.irridia.com 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:46:32 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T6000 arm64
% go version
go version go1.17.9 darwin/arm64
% buf --version
1.3.1
% md5 `which buf`
MD5 (/Users/ken/local/bin/buf) = 9543c69add0dddbfe90ae5da41b2d8aa
% xcodebuild -version 
Xcode 13.3.1
Build version 13E500a
% clang --version
Apple clang version 13.1.6 (clang-1316.0.21.2.3)
Target: arm64-apple-darwin21.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by nashikb** [comment]
@bufdev Sorry for the @, but this feels like a potentially widespread issue.

**Comment by seankimdev** [comment]
Hi @nashikb, let me relay this to our team internally and look into this.

**Comment by bufdev** [comment]
We will look into it. For what it’s worth, almost all of us use arm64
ourselves with the same setup, and can’t repro at the moment, and Buf is
generally more stable than the alternative. We do take this seriously
though and will investigate.

**Comment by bufdev** [comment]
Can you let us know the last version of Buf that did not have such issue
(we assume an upgrade caused an issue in your production workflows)? Can
you also provide exact repro steps? Does the bare minimum mean empty files?

**Comment by nashikb** [comment]
I found the issue. If you duplicate a proto package directory and re-run buf, it crashes. Fixing the first file processed by updating its package name will allow normal error reporting to occur. In other words, duplicate `package` specs cause a segfault.

Definitely "solved" from my perspective, and definitely user error, but the crash behavior makes it hard to understand what's going on. :-)

**Comment by bufdev** [comment]
Typically, reproduction steps will include exact steps to reproduce the bug - I can't follow how to reproduce the issue from your description above. [Here's an article](https://medium.com/we-are-testers/chapter-4-how-to-write-clear-and-efficient-steps-to-reproduce-in-your-bug-report-d96039a1c77d) covering how to produce good reproduction steps.

Fortunately, it appears as though the issue has been found in https://github.com/jhump/protoreflect/issues/508, and we will update `buf` once the fix for this is merged.

**Comment by bufdev** [comment]
For posterity: my interpretation of "duplicate package specs" is the following:

```proto
syntax = "proto3";

package a;
package a;
```

This properly errors when built with `buf`:

```
$ buf build
a.proto:4:1:files should have only one package declaration
```

I'm unable to deduce what the issue is from the above.

**Comment by bufdev** [comment]
I also tried this ("duplicate package names in subdirs"):

```
$ tree
.
├── a
│   └── a.proto
├── b
│   └── b.proto

$ cat a/a.proto
syntax = "proto3";

package a;

$ cat b/b.proto
syntax = "proto3";

package a;

$ buf build
# successful
```

So this is not what the author is referring to either.

**Comment by jhump** [comment]
The panic line is trying to report a symbol collision. That means there is a symbol -- like a message, service, etc -- with the same relative name. So putting the two files in the same package results in a fully-qualified name collision.

However, my suspicion is that it is unable to find the AST node for the symbol, which is why there is an unexpected nil interface (which is then de-referenced to call a method to get the file+line+column position). I'm not sure how that could happen, but one thought I had was with synthetic elements -- such as a map entry message or the synthetic oneof that wraps a "proto3 optional" field. Since they are synthetic, the AST node lookup may be failing and returning nil. I'll trace through a little more to see if I can figure out precisely what kind of collision might cause such an issue.

I can temporarily put a stop-gap in to have it report only the file name if the AST node lookup fails. So that would at least remedy the panic, but leave a bug where sometimes the line+column info is missing.

**Comment by jhump** [comment]
Here's the tracking issue in the protoreflect repo: jhump/protoreflect#508

**Comment by bufdev** [comment]
This has been fixed in v1.4.0.

**Comment by nashikb** [comment]
Sorry I didn't get back to this sooner with specific file examples; I went down a rabbit hole on an openapi-generator bug.

I can confirm that with 1.4.0 I now get the proper error messages and no crash. 👍
